File: packages/sdk-ts/src/adapters/embedding/onnx.ts
/**
 * ONNX all-MiniLM-L6-v2 embedding provider (optional `@huggingface/transformers`
 * peer dep).
 *
 * The REAL embedder (rec-embedding-port): the same
 * `sentence-transformers/all-MiniLM-L6-v2` ONNX artifact run by
 * `@huggingface/transformers` (transformers.js, TS) and `fastembed` (Py) —
 * parity by artifact, not reimplementation. Cosine ≈ 1 across the two runtimes.
 *
 * Lazy-download + cache, the Chroma pattern: the model is a downloaded artifact,
 * never a bundled dependency. `@huggingface/transformers` is dynamically
 * imported on the FIRST `embed` call and the pipeline fetches + caches the ONNX
 * then — so this module never pulls transformers.js into a process that does
 * not embed, and it is not in the default dependency graph.
 */
import type { EmbeddingPort } from "../../kernel/protocols.js";

// Same model + vector width as the Py twin (adapters/embedding/onnx.py) and the
// fake floor (FAKE_EMBEDDING_DIMS) so providers are swap-compatible.
export const ONNX_MODEL_ID = "Xenova/all-MiniLM-L6-v2";
export const ONNX_DIMS = 384;

/**
 * `EmbeddingPort` backed by transformers.js feature-extraction. The heavy
 * dynamic import (`@huggingface/transformers`) is deferred to first use so
 * merely importing this module never pulls ML deps into a process that does not
 * embed.
 */
export class OnnxEmbeddingProvider implements EmbeddingPort {
  readonly modelId: string;
  readonly dims: number;
  // The transformers.js pipeline, lazily constructed on first embed().
  private _pipe: unknown = null;

  constructor(modelId: string = ONNX_MODEL_ID, dims: number = ONNX_DIMS) {
    this.modelId = modelId;
    this.dims = dims;
  }

  private async ensurePipe(): Promise<
    (texts: string[], opts: { pooling: "mean"; normalize: boolean }) => Promise<{
      tolist(): number[][];
    }>
  > {
    if (this._pipe === null) {
      let mod: { pipeline: (task: string, model: string) => Promise<unknown> };
      // NON-LITERAL specifier on purpose: the peer dep is OPTIONAL and its
      // type declarations are NOT installed by default, so a literal
      // `import("@huggingface/transformers")` would fail `tsc --noEmit` in CI
      // with TS2307. A string variable makes the dynamic import resolve to
      // `Promise<any>` at type-check time (still the same module at runtime).
      const spec: string = "@huggingface/transformers";
      try {
        // Dynamic import so the dep is optional and never bundled by default.
        mod = (await import(/* @vite-ignore */ spec)) as typeof mod;
      } catch (err) {
        throw new Error(
          "OnnxEmbeddingProvider needs the '@huggingface/transformers' peer dep: "
            + "npm install @huggingface/transformers",
          { cause: err },
        );
      }
      // Downloads + caches the ONNX artifact on first construction.
      this._pipe = await mod.pipeline("feature-extraction", this.modelId);
    }
    return this._pipe as never;
  }

  async embed(texts: string[]): Promise<number[][]> {
    if (texts.length === 0) return [];
    const pipe = await this.ensurePipe();
    // Mean-pooled + L2-normalized sentence vectors — same recipe fastembed
    // uses by default, matching sentence-transformers.
    const out = await pipe(texts, { pooling: "mean", normalize: true });
    return out.tolist();
  }
}

File: packages/sdk-py/dna/adapters/embedding/onnx.py
"""ONNX all-MiniLM-L6-v2 embedding provider (opt-in ``embed-onnx`` extra).

The REAL embedder (rec-embedding-port): the same ``sentence-transformers/
all-MiniLM-L6-v2`` ONNX artifact run by ``fastembed`` (Py) and
``@huggingface/transformers`` (TS) — parity by artifact, not reimplementation.
Cosine ≈ 1 across the two runtimes (see ``tests/test_embedding_onnx_parity.py``,
which is network-gated and skips offline).

Lazy-download + cache, the Chroma pattern: the model is a downloaded artifact,
never an install-time dependency. ``fastembed`` fetches and caches the ONNX on
the FIRST ``embed`` call, not at import — so importing this module (with the
extra installed) still costs nothing until you actually embed.
"""
from __future__ import annotations

# Same model + vector width as the TS twin (src/adapters/embedding/onnx.ts) and
# the fake floor (FAKE_EMBEDDING_DIMS) so providers are swap-compatible.
ONNX_MODEL_ID = "sentence-transformers/all-MiniLM-L6-v2"
ONNX_DIMS = 384


class OnnxEmbeddingProvider:
    """``EmbeddingPort`` backed by fastembed's ONNX runtime.

    Structurally satisfies ``dna.kernel.protocols.EmbeddingPort``. The heavy
    import (``fastembed``, which drags ``onnxruntime``) is deferred to first
    use so merely importing this module never pulls ML deps into a process
    that does not embed."""

    def __init__(self, model_id: str = ONNX_MODEL_ID, dims: int = ONNX_DIMS) -> None:
        self.model_id = model_id
        self.dims = dims
        self._model = None  # lazily constructed on first embed()

    def _ensure_model(self):
        if self._model is None:
            try:
                from fastembed import TextEmbedding
            except ImportError as exc:  # pragma: no cover - exercised via extra
                raise ImportError(
                    "OnnxEmbeddingProvider needs the 'embed-onnx' extra: "
                    "pip install 'dna-sdk[embed-onnx]'"
                ) from exc
            # Downloads + caches the ONNX artifact on first construction.
            self._model = TextEmbedding(model_name=self.model_id)
        return self._model

    async def embed(self, texts: list[str]) -> list[list[float]]:
        if not texts:
            return []
        model = self._ensure_model()
        # fastembed yields L2-normalized mean-pooled sentence vectors (same
        # recipe transformers.js uses with pooling='mean', normalize=true).
        return [list(map(float, vec)) for vec in model.embed(list(texts))]
